ccl/kvccl/kvfollowerreadsccl: TestFollowerReadsWithStaleDescriptor failed

ccl/kvccl/kvfollowerreadsccl.TestFollowerReadsWithStaleDescriptor [failed](https://tanzanite.cluster.engflow.com/invocations/default/d6131356-c521-4ad1-a458-6674f2f5ecbb?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2t2Y2NsL2t2Zm9sbG93ZXJyZWFkc2NjbDprdmZvbGxvd2VycmVhZHNjY2xfdGVzdA==) on release-24.1 @ [c4641acfdb8533ccf951535f408ec61df8437b62](https://github.com/cockroachdb/cockroach/commits/c4641acfdb8533ccf951535f408ec61df8437b62):


```
=== RUN   TestFollowerReadsWithStaleDescriptor
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestFollowerReadsWithStaleDescriptor691115469
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go:891: (TestFollowerReadsWithStaleDescriptor)
	NOTICE: .DistSenderI() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().DistSenderI() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    followerreads_test.go:895: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go:895
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestFollowerReadsWithStaleDescriptor
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestFollowerReadsWithStaleDescriptor691115469
--- FAIL: TestFollowerReadsWithStaleDescriptor (91.21s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestFollowerReadsWithStaleDescriptor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39653

--- a/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
+++ b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
@@ -824,10 +824,12 @@ func TestFollowerReadsWithStaleDescriptor(t *testing.T) {
 	historicalQuery.Store(`SELECT * FROM test AS OF SYSTEM TIME follower_read_timestamp() WHERE k=2`)
 	recCh := make(chan tracingpb.Recording, 1)
 
+	settings := cluster.MakeClusterSettings()
 	tc := testcluster.StartTestCluster(t, 4,
 		base.TestClusterArgs{
 			ReplicationMode: base.ReplicationManual,
 			ServerArgs: base.TestServerArgs{
+				Settings:          settings,
 				DefaultTestTenant: base.TODOTestTenantDisabled,
 				UseDatabase:       "t",
 			},
@@ -865,6 +867,10 @@ func TestFollowerReadsWithStaleDescriptor(t *testing.T) {
 			},
 		})
 	defer tc.Stopper().Stop(ctx)
+
+	// Further down, we'll set up the test to pin the lease to store 1. Turn off
+	// load based rebalancing to make sure it doesn't move.
+	kvserver.LoadBasedRebalancingMode.Override(ctx, &settings.SV, kvserver.LBRebalancingOff)
 
 	n1 := sqlutils.MakeSQLRunner(tc.Conns[0])
 	n1.Exec(t, `CREATE DATABASE t`)
@@ -1078,6 +1084,9 @@ func TestSecondaryTenantFollowerReadsRouting(t *testing.T) {
 			systemSQL := sqlutils.MakeSQLRunner(tc.Conns[0])
 			systemSQL.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.target_duration = '0.1s'`)
 			systemSQL.Exec(t, `SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '0.1s'`)
+			// Disable the store rebalancer to make sure leases stay where they are;
+			// the test cares about this.
+			systemSQL.Exec(t, `SET CLUSTER SETTING kv.allocator.load_based_rebalancing = off`)
 
 			historicalQuery := `SELECT * FROM t.test AS OF SYSTEM TIME follower_read_timestamp() WHERE k=2`
 			useExplainAnalyze := rng.Float64() < 0.5

